### ISSUE TITLE
X3DAUDIO_CALCULATE_ZEROCENTER should be set for proper volume levels returned by X3DAudioCalculate

### ISSUE THREAD

**Issue by HSNB** [microsoft/DirectXTK#521]
It seems that even when there's no Front Center speaker according to the Mastering Voice Channel Mask (SPEAKER_FRONT_CENTER), X3DAudioCalculate will still calculate the matrix volume levels as if there is.

This causes the volume of the sound to play at a lower volume.

For example if you have a 2 channel output (headset), meaning Channel Mask is SPEAKER_FRONT_LEFT | SPEAKER_FRONT_RIGHT. It is still needed to set X3DAUDIO_CALCULATE_ZEROCENTER even though the comment of X3DAUDIO_CALCULATE_ZEROCENTER says "valid only for matrix calculations with final mix formats that have a front center channel".

I could be wrong in understanding 3D audio, but if I'm not setting this ZEROCENTER flag what happens is that even a 3D sound playing at a distance of 0 away from the listener, makes the sound play at a lower volume than the actual sound file, which seems like a wrong thing to me?

**Comment by walbourn** [comment]
Interesting. Thanks for the feedback.

Looking at the source to XAudio3DCalculate, they clearly have validation that enforces the flag as documented:

```cpp
            if (u32Flags & X3DAUDIO_CALCULATE_ZEROCENTER) {
                if (!(u32Flags & X3DAUDIO_CALCULATE_MATRIX) || !(pHandle->u32SpeakerChannelMask & SPEAKER_FRONT_CENTER)) {
                    X3DDEBUG_MESSAGE(X3DDEBUG_ERROR, "X3DAUDIO_CALCULATE_ZEROCENTER valid only for matrix calculations with final mix formats that have a front center channel");
                    fValid = FALSE;
                }
            }
```

So if there is no FRONT_CENTER in the provided channel mask or you don't have the flag to calculate a matrix set, the debugging logic would say it's not valid.

Furthermore, the actual implementation for `X3DAUDIO_CALCULATE_ZEROCENTER` uses this flag specifically for the following speaker configs to remove one speaker channel:

```
SPEAKER_SURROUND
SPEAKER_5POINT1
SPEAKER_7POINT1
SPEAKER_5POINT1_SURROUND
SPEAKER_7POINT1_SURROUND
```

The other logic for that flag only applies if there are more than 2 speaker channels (since the first two are always left/right and the third is the center).

In other words, I don't see where in the code that flag would impact a 2 channel speaker configuration.

**Comment by walbourn** [comment]
Note that with that said, perhaps I should offer a way to set that flag through *DirectX Tool Kit for Audio* for scenarios where it is desireable.

**Comment by walbourn** [comment]
Can you provide the full list of values you are passing to `X3DAUDIO_DSP_SETTINGS` as well as the `SpeakerChannelMask` that was initially passed to `X3DAudioInitialize`?

Also make sure you are using the handle that came back from `X3DAudioInitialize` and not some other memory location.

**Comment by walbourn** [comment]
On my system, I have a stereo output device, a mono input sound, and the volume is the same matrix values are coming out of the calculation with or without `X3DAUDIO_CALCULATE_ZEROCENTER` set.

That said, I am adding new control flags to provide better control over the calculate flags used by the engine:

* Adds new audio engine flags
  * `AudioEngine_DisableLFERedirect` will disable the current default behavior of using `X3DAUDIO_CALCULATE_REDIRECT_TO_LFE` if there is a LFE channel in the output.
  * `AudioEngine_ZeroCenter3D` will add `X3DAUDIO_CALCULATE_ZEROCENTER` to all X3DAudio calculations.

* Adds new sound instance flags
  * `SoundEffectInstance_UseRedirectLFE` internal flag now repurposed to force use of `X3DAUDIO_CALCULATE_REDIRECT_TO_LFE` per voice if not enabled globally and if there is a LFE channel in the output.
  * `SoundEffectInstance_ZeroCenter3D` flag forces the use of  `X3DAUDIO_CALCULATE_ZEROCENTER` per voice if not enabled globally.

> From what I see in the X3DAudio code in Windows, I would expect that I should disable the use of `X3DAUDIO_CALCULATE_ZEROCENTER` if there is no center speaker in the configuration, but I left that 'guard' out to help your scenario.

**Comment by walbourn** [comment]
https://github.com/microsoft/DirectXTK/pull/543

https://github.com/microsoft/DirectXTK12/pull/312
